# Case Study on Online Movie Rental Company

This is based on a case study about an online movie rental company with a database about customer information, movie ratings, background information on actors and more. 

Applied SQL queries to study for example customer preferences, customer engagement, and sales development and also used SQL extensions for online analytical processing (OLAP), which makes it easier to obtain key insights from multidimensional aggregated data.

# Introduction

Worked with a Postgres database from a fictional movie rental company called MovieNow. 

MovieNow offers an online platform for streaming movies. Customers can rent a movie for 24 hours. For all movies, the company stores additional information such as the genre or the main actors. MovieNow also stores information about customers and movie ratings.


# Schema

The schema contains 5 tables : customers, movies, renting, actors, actsin.

Customers (customer_id, name, country, gender, date of birth, date_account_start) 
 
Movies (movie_id, title, genre, runtime, year_of_releas, renting_price) 

Renting (renting_id, customer_id, movie_id, rating, date_renting) 

Actors (actor_id, name, year_of_birth, nationality, gender) 

Actsin (actsin_id, actor_id, movie_id) 

# KPI's

KPI's help company define and monitor success. For the MovieNow Company KPI's are:

Revenue - the total number of rentals.

Customer satisfaction - average rating of all movies.

Customer Engagement - number of active customers.

# Data Exploration

In [1]:
SELECT *
FROM movies.actors
LIMIT 5;

actor_id                name  year_of_birth nationality  gender
0         1       Abbie Cornish           1982   Australia  female
1         2        Adam Sandler           1966         USA    male
2         3           Al Pacino           1940         USA    male
3         4           Amy Adams           1974         USA  female
4         5  Andrea Riseborough           1981     British  female

Let's see how many actors are from 'USA'

In [10]:
SELECT COUNT(*)
FROM movies.actors
WHERE nationality = 'USA';

count
0     91

There are 91 actors from USA.

In [6]:
SELECT *
FROM movies.customers
LIMIT 5;

customer_id  ...        date_account_start
0            2  ... 2018-10-15 00:00:00+00:00
1            3  ... 2019-02-14 00:00:00+00:00
2            4  ... 2017-11-22 00:00:00+00:00
3            5  ... 2018-12-16 00:00:00+00:00
4            6  ... 2018-08-26 00:00:00+00:00

[5 rows x 6 columns]

Lets see the number of countries where MovieNow has customers.

In [11]:
SELECT COUNT( DISTINCT country)
FROM movies.customers;

count
0     11

MovieNow has customers in 11 countries 

In [3]:
SELECT *
FROM movies.movies
LIMIT 5;

movie_id                       title  ... year_of_release  renting_price
0         1       One Night at McCool's  ...            2001           2.09
1         2                   Swordfish  ...            2001           2.19
2         3             What Women Want  ...            2001           2.59
3         4                Training Day  ...            2001           1.79
4         5  The Fellowship of the Ring  ...            2001           2.59

[5 rows x 6 columns]

Let's see the distinct available genres 

In [17]:
SELECT DISTINCT genre
FROM movies.movies;

genre
0                  Animation
1                      Other
2                      Drama
3  Art House & International
4         Mystery & Suspense
5                     Comedy
6         Action & Adventure
7  Science Fiction & Fantasy

There are 8 distinct genres in the database.

In [4]:
SELECT *
FROM movies.actsin
LIMIT 5;

actsin_id  movie_id  actor_id
0          1        37         1
1          2        56         2
2          3        10         3
3          4        14         3
4          5        29         3

In [5]:
SELECT *
FROM movies.renting
LIMIT 5;

renting_id  customer_id  movie_id  rating              date_renting
0           1           41         8     NaN 2018-10-09 00:00:00+00:00
1           2           10        29    10.0 2017-03-01 00:00:00+00:00
2           3          108        45     4.0 2018-06-08 00:00:00+00:00
3           4           39        66     8.0 2018-10-22 00:00:00+00:00
4           5          104        15     7.0 2019-03-18 00:00:00+00:00

Lets calc. the avg. rating per movie.

In [6]:
SELECT movie_id, rating
FROM movies.renting;

movie_id  rating
0           8     NaN
1          29    10.0
2          45     4.0
3          66     8.0
4          15     7.0
..        ...     ...
573        14     NaN
574        31    10.0
575        25     NaN
576        65     9.0
577        16     NaN

[578 rows x 2 columns]

Looks like the column rating in the renting table has a lot of null values since not every customer gives a rating after watching a movie.

# Data Analysis 

## 1. First account for each country

In [1]:
-- Finding when the first customer accounts were created for each country

SELECT country, MIN(date_account_start) AS first_account
FROM movies.customers
GROUP BY country
ORDER BY first_account;


country             first_account
0         France 2017-01-13 00:00:00+00:00
1        Hungary 2017-01-18 00:00:00+00:00
2        Belgium 2017-01-28 00:00:00+00:00
3       Slovenia 2017-01-31 00:00:00+00:00
4          Spain 2017-02-14 00:00:00+00:00
5          Italy 2017-02-28 00:00:00+00:00
6         Poland 2017-03-03 00:00:00+00:00
7   Great Britan 2017-03-31 00:00:00+00:00
8        Denmark 2017-04-30 00:00:00+00:00
9            USA 2017-09-13 00:00:00+00:00
10       Austria 2017-11-22 00:00:00+00:00

We can see that the first customer account was created in France.

## 2. Average movie ratings

In [2]:
-- lets see for each movie the average rating, the number of ratings and the number of views

SELECT movie_id, 
       AVG(rating) AS avg_rating,
       COUNT(rating) AS number_ratings,
       COUNT(*) AS number_renting
FROM movies.renting
GROUP BY movie_id
ORDER BY avg_rating DESC; 

movie_id  avg_rating  number_ratings  number_renting
0         11         NaN               0               6
1         48   10.000000               2               5
2          5    9.750000               4               8
3         42    9.600000               5              10
4         63    9.333333               6              11
..       ...         ...             ...             ...
66        14    6.500000               6              11
67        68    6.333333               3               7
68        59    6.000000               2               5
69        43    6.000000               1               4
70        56    5.666667               3               4

[71 rows x 4 columns]

The average rating for movie_id 11  is null because all rating values are null for that movie id.

## 3. Average rating per customer

In [1]:
-- We see the avg. movie ratings given by each customer, number of ratings, number of rentals and analyse customers having rentals more than 7.

SELECT customer_id, 
       AVG(rating) AS avg_rating,
       COUNT(rating) AS number_ratings,
       COUNT(*) AS number_renting
FROM movies.renting
GROUP BY customer_id
HAVING COUNT(*)>7
ORDER BY avg_rating; 

customer_id  avg_rating  number_ratings  number_renting
0           104    6.250000               4               8
1            28    6.714286               7              11
2           111    7.000000               3              10
3           113    7.000000               7              15
4            25    7.200000               5              10
5            21    7.333333               6              14
6            92    7.571429               7              11
7            49    7.625000               8              13
8            35    7.666667               6               9
9            52    7.875000               8               9
10          108    8.000000               6              10
11          114    8.000000               6              11
12           66    8.000000               7              10
13           93    8.400000               5               8
14           26    8.400000               5              10
15           94    8.714286               7              10
16          121    8.750000               4               9
17            7    9.000000               4               9
18           61    9.200000               5               8
19            8    9.250000               4               8
20           99    9.250000               4               8
21           65    9.333333               3               9

We see the customer number 104 gave the lowest average ratings for 4 movies.

## 4. Joining Renting and Customers.

In [5]:
-- Adding customer info to renting table to find the avg. ratings of customers from France

SELECT AVG(rating) 
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
WHERE c.country='France';

avg
0  7.771429

The average rating for customers from France is 7.77.

## 5. Aggregating revenue, rentals and active customers.

In [6]:
-- Reporting the KPI's for the performance of company in 2018.

SELECT 
	SUM(m.renting_price) AS revenue, 
	COUNT(*) AS number_rentals, 
	COUNT(DISTINCT r.customer_id) AS number_active_customers
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON r.movie_id = m.movie_id
WHERE date_renting BETWEEN '2018-01-01' AND '2018-12-31' ;

revenue  number_rentals  number_active_customers
0   658.02             298                       93

We calculated a turnover of 658.02 and found the number of rentals to be 298 and the number of active users to be 93 in 2018.

## 6. Movies and Actors

In [7]:
--Listing which actors played in which movie

SELECT m.title, a.name
FROM movies.actsin AS ai
LEFT JOIN movies.movies AS m
ON m.movie_id = ai.movie_id
LEFT JOIN movies.actors AS a
ON a.actor_id = ai.actor_id

title              name
0                               Candy     Abbie Cornish
1                       Jack and Jill      Adam Sandler
2                              Simone         Al Pacino
3                         The Recruit         Al Pacino
4                   Two for the Money         Al Pacino
..                                ...               ...
208                          The Help       Viola Davis
209  Ghost Rider: Spirit of Vengeance  Violante Placido
210                             Rio 2         will.i.am
211                      The Sessions   William H. Macy
212                 Failure to Launch   Zooey Deschanel

[213 rows x 2 columns]

## 7. Income from movies.

In [1]:
SELECT rm.title,
SUM(rm.renting_price) AS income_movie
FROM
       (SELECT m.title,  
               m.renting_price
       FROM movies.renting AS r
       LEFT JOIN movies.movies AS m
       ON r.movie_id=m.movie_id) AS rm
GROUP BY rm.title
ORDER BY income_movie DESC; 

title  income_movie
0        Bridget Jones - The Edge of Reason         37.57
1                                 Fair Game         34.68
2                               The Kingdom         31.35
3                         Two for the Money         30.69
4                          Django Unchained         29.59
..                                      ...           ...
66  Harry Potter and the Chamber of Secrets          8.95
67                            Jack and Jill          8.36
68                             The Invasion          8.36
69                          The Human Stain          7.96
70                          The Other Woman          3.38

[71 rows x 2 columns]

Bridget Jones movies has the highest income of 37.57

## 8.Age of actors from the USA

In [2]:
-- Explore the age of American actor and actress.

SELECT a.gender,MIN(a.year_of_birth), MAX(a.year_of_birth)
FROM (SELECT * FROM movies.actors  WHERE nationality = 'USA') as a
GROUP BY a.gender;

gender   min   max
0  female  1945  1993
1    male  1930  1992

MovieNow actors' record the oldest actor was born in 1930 and the oldest actress in 1945.

## 9. Identifying favorite movies for a group of customers

In [4]:
-- Lets take the case for a specific group of customers: for all customers born in the 70s.

SELECT m.title, 
COUNT(*),
AVG(r.rating)
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31'
GROUP BY m.title
HAVING COUNT(*)>1 -- Remove movies with only one rental
ORDER BY AVG(r.rating) DESC; -- Order with highest rating first

title  count        avg
0                                        Showtime      5        NaN
1   Harry Potter and the Deathly Hallows – Part 2      2        NaN
2                               Waking Up in Reno      2        NaN
3                                             Ray      2        NaN
4                                Django Unchained      4  10.000000
5                                     The Fighter      4  10.000000
6                           One Night at McCool's      2  10.000000
7                          No Country for Old Men      3  10.000000
8                      The Fellowship of the Ring      2  10.000000
9                                   I'm Not There      2  10.000000
10                               Secondhand Lions      4   9.666667
11                                  Winter's Bone      2   9.500000
12                                    Fool's Gold      4   9.333333
13                                      25th Hour      2   9.000000
14                             Burn After Reading      2   9.000000
15                                  Morning Glory      2   9.000000
16                                       The Help      3   9.000000
17                              Failure to Launch      4   8.666667
18                                    Young Adult      4   8.500000
19                                   Sleepwalking      2   8.500000
20                                    Black Death      3   8.500000
21             Bridget Jones - The Edge of Reason      6   8.250000
22                           Welcome to Mooseport      4   8.000000
23         Harry Potter and the Half-Blood Prince      3   8.000000
24                           The Royal Tenenbaums      2   8.000000
25                                      Fair Game      3   8.000000
26                                        Monster      5   8.000000
27                              The Frozen Ground      4   7.500000
28       Harry Potter and the Prisoner of Azkaban      4   7.500000
29      Harry Potter and the Order of the Phoenix      3   7.500000
30                              Two for the Money      4   7.333333
31                                   A Good Woman      4   7.333333
32                                   Training Day      3   7.000000
33                                          Candy      3   7.000000
34                                         Capote      3   7.000000
35                                 Shall We Dance      4   7.000000
36                                     Happy Feet      3   7.000000
37                                 V for Vendetta      2   7.000000
38                                 The Two Towers      3   7.000000
39                                    The Kingdom      5   6.666667
40                               The Hunger Games      2   6.500000
41                             World Trade Center      3   6.500000
42                                         Simone      5   6.000000
43                              Then She Found Me      4   6.000000
44  Harry Potter and the Deathly Hallows – Part 1      2   6.000000
45                         Something's Gotta Give      3   6.000000
46                                    The Recruit      2   5.000000

'Django Unchained' is a movie with one of the best ratings among people born in the 70s.

## 10. Identify favorite actors for Spain

In [5]:
SELECT a.name,  c.gender,
       COUNT(*) AS number_views, 
       AVG(r.rating) AS avg_rating
FROM movies.renting as r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
LEFT JOIN movies.actsin as ai
ON r.movie_id = ai.movie_id
LEFT JOIN movies.actors as a
ON ai.actor_id = a.actor_id
WHERE c.country = 'Spain' -- Select only customers from Spain
GROUP BY a.name, c.gender
HAVING AVG(r.rating) IS NOT NULL 
  AND COUNT(*) > 5 
ORDER BY avg_rating DESC, number_views DESC;

name  gender  number_views  avg_rating
0  Catherine Keener  female             6         8.0
1       Emma Watson    male             7         7.6
2  Daniel Radcliffe    male             7         7.6
3      Rupert Grint    male             7         7.6

Catherine Keener is the favorite actress among female Spain customers.

Male customers from Spain like the actors from Harry Potter best: Emma Watson, Daniel Radcliffe and Rupert Grint.

## 11. KPI's per country

In [6]:
-- Reporting KPI's for each country from begining of year 2019.

SELECT 
	 c.country ,               
	COUNT(*) AS number_renting, 
	AVG(r.rating) AS average_rating, 
	SUM(m.renting_price) AS revenue        
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE date_renting >= '2019-01-01'
GROUP BY c.country;


country  number_renting  average_rating  revenue
0           None               1       10.000000     1.79
1          Spain              26        8.076923    57.94
2   Great Britan               9        7.200000    17.91
3        Austria               4        6.000000     8.16
4          Italy              30        8.235294    69.90
5         Poland              21        8.125000    49.29
6       Slovenia              14        8.000000    30.26
7        Hungary              11        7.166667    25.89
8        Denmark               7        7.714286    14.63
9        Belgium               7        9.000000    13.13
10           USA               6        9.000000    14.54
11        France              23        7.533333    51.07

We calculated a total revenue of 57.94 for Spain, with 26 movie rentals and an average rating of 8.1.

## 12. Often rented movies.

In [1]:
-- List all movies with more than 5 views

SELECT *
FROM movies.movies
WHERE movie_id IN  
(SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(*) > 5)

movie_id  ... renting_price
0          1  ...          2.09
1          2  ...          2.19
2          3  ...          2.59
3          4  ...          1.79
4          5  ...          2.59
5          6  ...          2.69
6          7  ...          1.89
7         10  ...          2.69
8         11  ...          1.79
9         12  ...          2.39
10        14  ...          1.69
11        15  ...          1.69
12        16  ...          1.59
13        18  ...          1.69
14        19  ...          2.09
15        22  ...          2.59
16        23  ...          1.69
17        24  ...          2.89
18        25  ...          2.89
19        26  ...          2.59
20        27  ...          2.09
21        28  ...          1.69
22        29  ...          2.79
23        30  ...          2.39
24        31  ...          2.69
25        32  ...          2.89
26        33  ...          1.59
27        34  ...          1.99
28        35  ...          1.79
29        36  ...          1.59
30        37  ...          1.59
31        38  ...          2.49
32        39  ...          1.79
33        40  ...          1.99
34        41  ...          2.09
35        42  ...          1.49
36        44  ...          2.49
37        45  ...          2.39
38        46  ...          1.79
39        47  ...          2.69
40        49  ...          1.59
41        50  ...          2.69
42        51  ...          2.99
43        52  ...          2.89
44        53  ...          2.49
45        54  ...          2.79
46        57  ...          1.99
47        58  ...          2.49
48        61  ...          1.99
49        62  ...          2.29
50        63  ...          2.69
51        64  ...          2.89
52        66  ...          1.59
53        67  ...          1.59
54        68  ...          2.39
55        69  ...          2.49
56        70  ...          1.59
57        71  ...          1.79

[58 rows x 6 columns]

58 movies are often rented in which 7 belong to Harry Potter Series

## 13. Frequent Customers

In [2]:
-- List all info of customers who rented more than 10 movies.

SELECT *
FROM movies.customers
WHERE customer_id IN           
	(SELECT customer_id
	FROM movies.renting
	GROUP BY customer_id
	HAVING COUNT(*)>10);

customer_id  ...        date_account_start
0           21  ... 2017-06-11 00:00:00+00:00
1           28  ... 2017-02-04 00:00:00+00:00
2           49  ... 2017-01-18 00:00:00+00:00
3           92  ... 2017-09-21 00:00:00+00:00
4          113  ... 2017-06-13 00:00:00+00:00
5          114  ... 2017-02-14 00:00:00+00:00

[6 rows x 6 columns]

Avelaine Corbeil is one of the frequent customer from France

## 14. Movies with rating above average

In [3]:
SELECT title 
FROM movies.movies
WHERE movie_id IN
	(SELECT movie_id
	 FROM movies.renting
     GROUP BY movie_id
     HAVING AVG(rating) > 
		(SELECT AVG(rating)
		 FROM movies.renting));

title
0                                 What Women Want
1                      The Fellowship of the Ring
2        Harry Potter and the Philosopher's Stone
3                            The Royal Tenenbaums
4                               Waking Up in Reno
5                          11'09''01 September 11
6         Harry Potter and the Chamber of Secrets
7                                       25th Hour
8                                 The Human Stain
9                                Secondhand Lions
10                            Imagining Argentina
11                                 Shall We Dance
12                           Welcome to Mooseport
13                                            Ray
14                                        Monster
15       Harry Potter and the Prisoner of Azkaban
16                              Two for the Money
17                                 V for Vendetta
18                              Failure to Launch
19                                     Happy Feet
20                                 The Lake House
21                                  I'm Not There
22                              Then She Found Me
23                         No Country for Old Men
24                                    Fool's Gold
25                                      Astro Boy
26         Harry Potter and the Half-Blood Prince
27                                  Winter's Bone
28                                  Morning Glory
29                                    The Fighter
30  Harry Potter and the Deathly Hallows – Part 1
31                                    Young Adult
32                                       The Help
33                                    Black Death
34  Harry Potter and the Deathly Hallows – Part 2
35                                   The Sessions
36                               Django Unchained
37                                    Cloud Atlas
38               Ghost Rider: Spirit of Vengeance
39                                          Rio 2

Movies 'What Women Want' and 'The Fellowship of the Ring' are having an average rating higher than the total average and also the most often rented movies.

## 15. Analyzing customer behaviour

In [1]:
-- Lets focus on customers who rented fewer than 5 movies.

SELECT *
FROM movies.customers as c
WHERE 5 > 
	(SELECT count(*)
	FROM movies.renting as r
	WHERE r.customer_id = c.customer_id);

customer_id  ...        date_account_start
0             2  ... 2018-10-15 00:00:00+00:00
1             3  ... 2019-02-14 00:00:00+00:00
2             4  ... 2017-11-22 00:00:00+00:00
3             5  ... 2018-12-16 00:00:00+00:00
4             6  ... 2018-08-26 00:00:00+00:00
..          ...  ...                       ...
61          117  ... 2018-11-09 00:00:00+00:00
62          119  ... 2019-03-06 00:00:00+00:00
63          120  ... 2018-08-06 00:00:00+00:00
64          122  ... 2018-11-27 00:00:00+00:00
65          123  ... 2018-12-30 00:00:00+00:00

[66 rows x 6 columns]

## 16. Customers who gave lower ratings

In [2]:
-- Let's identify customers who were not satisfied with movies they watched on MovieNow and report a list of customers with minimum rating smaller than 4.

SELECT *
FROM movies.customers AS c
WHERE 4 >  
	(SELECT MIN(rating)
	FROM movies.renting AS r
	WHERE r.customer_id = c.customer_id);


customer_id  ...        date_account_start
0           28  ... 2017-02-04 00:00:00+00:00
1           41  ... 2017-06-12 00:00:00+00:00
2           86  ... 2017-12-15 00:00:00+00:00
3          120  ... 2018-08-06 00:00:00+00:00

[4 rows x 6 columns]

Customers Sidney Généreux, Zara Mitchell, Albin Jaworski, and Robin J. Himes rated a movie with less than 4.

## 17. Most attention received movies on platform

In [3]:
-- report all movies with more than 5 ratings and all movies with an average rating higher than 8.

SELECT *
FROM movies.movies as m
WHERE 5 < 
	(SELECT COUNT(rating)
	FROM movies.renting as r
	WHERE r.movie_id = m.movie_id)
AND 8 <
	(SELECT AVG(rating)
	FROM movies.renting AS r
	WHERE r.movie_id = m.movie_id);

movie_id  ... renting_price
0        22  ...          2.59
1        27  ...          2.09
2        40  ...          1.99
3        49  ...          1.59
4        51  ...          2.99
5        54  ...          2.79
6        57  ...          1.99
7        63  ...          2.69

[8 rows x 6 columns]

We see Django Unchained is one of the movies that has more than 5 ratings with avg. rating higher than 8.

## 18. Customer with at least one rating

In [4]:
SELECT *
FROM movies.customers as c 
WHERE EXISTS
	(SELECT *
	FROM movies.renting AS r
	WHERE rating IS NOT NULL 
	AND r.customer_id = c.customer_id);

customer_id  ...        date_account_start
0              2  ... 2018-10-15 00:00:00+00:00
1              4  ... 2017-11-22 00:00:00+00:00
2              7  ... 2018-05-12 00:00:00+00:00
3              8  ... 2018-02-08 00:00:00+00:00
4             10  ... 2017-01-28 00:00:00+00:00
..           ...  ...                       ...
98           118  ... 2017-11-24 00:00:00+00:00
99           119  ... 2019-03-06 00:00:00+00:00
100          120  ... 2018-08-06 00:00:00+00:00
101          121  ... 2017-09-13 00:00:00+00:00
102          123  ... 2018-12-30 00:00:00+00:00

[103 rows x 6 columns]

Looks like there are 19 customers who haven't given any ratings.

## 19. Actors in comedies

In [7]:
-- list of actors who play in comedies and then, the number of actors for each nationality playing in comedies.

SELECT  a.nationality,COUNT(*)
FROM movies.actors AS a
WHERE EXISTS
	(SELECT ai.actor_id
	 FROM movies.actsin AS ai
	 LEFT JOIN movies.movies AS m
	 ON m.movie_id = ai.movie_id
	 WHERE m.genre = 'Comedy'
	 AND ai.actor_id = a.actor_id)
GROUP BY a.nationality;

nationality  count
0  Northern Ireland      1
1               USA     22
2      South Africa      1
3            Canada      1
4           British      3

We see there is one actor each coming from South Africa, Canada and Northen Ireland, three actors from Great Britain, and 22 from the USA who played in a Comedy.

## 20. Young actors not coming from the USA

In [11]:
-- Identify actors who are not from the USA and actors who were born after 1990.

SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE nationality <> 'USA'
INTERSECT 
SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE year_of_birth > 1990;

name nationality  year_of_birth
0  Freddie Highmore     British           1992

Freddie Highmore as the only actor who is not from the USA and born after 1990.

## 21. Dramas with high ratings


In [12]:
-- The advertising team has a new focus. They want to draw the attention of the customers to dramas. So list of all movies that are in the drama genre and have an average rating higher than 9.

SELECT *
FROM movies.movies
WHERE movie_id IN 
   (SELECT movie_id
    FROM movies.movies
    WHERE genre = 'Drama'
    INTERSECT
    SELECT movie_id
    FROM movies.renting
    GROUP BY movie_id
    HAVING AVG(rating)>9);


movie_id                   title  ... year_of_release  renting_price
0        42  No Country for Old Men  ...            2007           1.49

[1 rows x 6 columns]

'No Country for Old Men' is the only drama with average rating higher than 9.

# OLAP - Online Analytical Processing

## 1. Group of Customers (CUBE) VS ROLLUP

In [1]:
-- Table with the total number of male and female customers from each country.

SELECT gender, 
	   country,
	   COUNT(*)
FROM movies.customers
GROUP BY CUBE (country,gender)
ORDER BY country;

gender       country  count
0   female       Austria      3
1     male       Austria      1
2     None       Austria      4
3     male       Belgium      3
4     None       Belgium      6
5   female       Belgium      3
6     None       Denmark      7
7   female       Denmark      4
8     male       Denmark      3
9     male        France      8
10    None        France     13
11  female        France      5
12    None  Great Britan     10
13  female  Great Britan      5
14    male  Great Britan      5
15    male       Hungary      6
16  female       Hungary      5
17    None       Hungary     11
18    None         Italy     27
19  female         Italy     14
20    male         Italy     13
21  female        Poland      4
22    male        Poland      9
23    None        Poland     13
24  female      Slovenia      7
25    male      Slovenia      4
26    None      Slovenia     11
27    male         Spain      6
28  female         Spain     10
29    None         Spain     16
30  female           USA      2
31    male           USA      2
32    None           USA      4
33    male          None     60
34  female          None     62
35    None          None    122

There are a total of 122 customers and that three out of four customers from Austria are female.

In [14]:
-- Look the total number of customers, the number of customers for each country, and the number of female and male customers for each country.

SELECT country,
       gender,
	   COUNT(*)
FROM movies.customers
GROUP BY  ROLLUP (country, gender)
ORDER BY country,gender;

country  gender  count
0        Austria  female      3
1        Austria    male      1
2        Austria    None      4
3        Belgium  female      3
4        Belgium    male      3
5        Belgium    None      6
6        Denmark  female      4
7        Denmark    male      3
8        Denmark    None      7
9         France  female      5
10        France    male      8
11        France    None     13
12  Great Britan  female      5
13  Great Britan    male      5
14  Great Britan    None     10
15       Hungary  female      5
16       Hungary    male      6
17       Hungary    None     11
18         Italy  female     14
19         Italy    male     13
20         Italy    None     27
21        Poland  female      4
22        Poland    male      9
23        Poland    None     13
24      Slovenia  female      7
25      Slovenia    male      4
26      Slovenia    None     11
27         Spain  female     10
28         Spain    male      6
29         Spain    None     16
30           USA  female      2
31           USA    male      2
32           USA    None      4
33          None    None    122

USA and Austria have the least number of customers of 4.

## 2. Categories of movies

In [2]:
-- number of movies for different genres and release years.

SELECT genre,
       year_of_release,
       COUNT(*)
FROM movies.movies
GROUP BY CUBE(genre,year_of_release)
ORDER BY year_of_release;

genre  year_of_release  count
0                        None           2001.0      6
1                       Drama           2001.0      2
2                      Comedy           2001.0      2
3   Science Fiction & Fantasy           2001.0      2
4                      Comedy           2002.0      3
..                        ...              ...    ...
57  Art House & International              NaN      1
58                      Drama              NaN     36
59                      Other              NaN      2
60                  Animation              NaN      2
61  Science Fiction & Fantasy              NaN     11

[62 rows x 3 columns]

Only one movie from 2014 is available on MovieNow. The highest number of movies is from 2003 with 8 movies.

## 3. Analyzing average ratings

In [4]:
-- the average rating of movies across countries and genres.

SELECT 
	c.country, 
	m.genre, 
	AVG(r.rating) AS avg_rating 
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
GROUP BY CUBE(c.country,m.genre); 

country                      genre  avg_rating
0       None                       None    7.939024
1     France         Mystery & Suspense    6.000000
2   Slovenia         Action & Adventure         NaN
3      Spain                  Animation         NaN
4     Poland                     Comedy    8.200000
..       ...                        ...         ...
91      None  Art House & International    8.666667
92      None         Mystery & Suspense    6.833333
93      None                     Comedy    7.766667
94      None         Action & Adventure    9.090909
95      None  Science Fiction & Fantasy    8.163934

[96 rows x 3 columns]

The average over all records is 7.94.

## 4. Analyzing preferences of genres across countries

In [11]:
-- Evaluate the preferences of customers by averaging their ratings and counting the number of movies rented from each genre.

SELECT 
	c.country, 
	m.genre, 
	AVG(rating) AS avg_rating, 
	COUNT(*) AS num_rating
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
GROUP BY ROLLUP (c.country,m.genre)
ORDER BY c.country, m.genre;

country                      genre  avg_rating  num_rating
0   Austria                     Comedy    8.000000           1
1   Austria                      Drama    6.000000           2
2   Austria         Mystery & Suspense         NaN           1
3   Austria  Science Fiction & Fantasy    6.666667           3
4   Austria                       None    6.800000           7
..      ...                        ...         ...         ...
83     None                     Comedy         NaN           1
84     None                      Drama    6.000000           2
85     None  Science Fiction & Fantasy         NaN           1
86     None                       None    7.939024         578
87     None                       None    8.000000           5

[88 rows x 4 columns]

Ther are total 578 ratings with Austria having least number of ratings of 7.

## 5. Exploring nationality and gender of actors

In [12]:
-- Count the number of actors in the table actors from each country, the number of male and female actors and the total number of actors.

SELECT 
	nationality ,
    gender ,
    COUNT(*)
FROM movies.actors
GROUP BY GROUPING SETS ((nationality), (gender), ()); 

nationality  gender  count
0               None    None    145
1          Argentina    None      1
2              Spain    None      3
3              Italy    None      1
4        Puerto Rico    None      1
5               Iran    None      1
6   Northern Ireland    None      2
7                USA    None     91
8             Israel    None      1
9        Netherlands    None      1
10      South Africa    None      1
11                      None      2
12           Austria    None      1
13         Australia    None      8
14           Ireland    None      2
15            Canada    None      4
16           British    None     24
17           Somalia    None      1
18              None    male     90
19              None  female     55

There are 91 out of 145 actors are from the USA and that there are 90 male and 55 female actors.

## 6. Exploring rating by country and gender


In [13]:
SELECT 
	c.country, 
    c.gender,
	AVG(r.rating)
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
GROUP BY GROUPING SETS ((country, gender), (country),(gender),());

country  gender       avg
0           None    None  7.939024
1        Austria    male  6.000000
2         France  female  8.000000
3        Hungary  female  7.285714
4           None    None  8.000000
5          Spain  female  7.612903
6        Belgium    male  8.000000
7            USA    male  7.500000
8        Denmark  female  8.444444
9        Austria  female  7.000000
10      Slovenia    male  8.090909
11       Belgium  female  9.125000
12      Slovenia  female  8.545455
13         Italy  female  8.366667
14       Denmark    male  7.333333
15  Great Britan    male  7.714286
16        France    male  7.666667
17        Poland  female  7.866667
18         Spain    male  7.681818
19         Italy    male  7.892857
20           USA  female  8.750000
21  Great Britan  female  7.272727
22       Hungary    male  7.857143
23        Poland    male  8.333333
24          None    None  8.000000
25         Spain    None  7.641509
26  Great Britan    None  7.520000
27       Austria    None  6.800000
28        Poland    None  8.121212
29         Italy    None  8.137931
30      Slovenia    None  8.431818
31       Hungary    None  7.628571
32       Denmark    None  7.888889
33       Belgium    None  8.900000
34        France    None  7.771429
35           USA    None  8.000000
36          None    None  8.000000
37          None  female  8.052941
38          None    male  7.814103

The overall average of ratings is 7.9.

# Business Case

MovieNow considers to invest money in new movies. Its generally more expensive to make movies available which were recently produced than older ones.

Check:

Do customers give better ratings to movies which were recently produced than older ones and is there any difference across countries.

In [16]:
SELECT c.country,
	m.year_of_release,
	COUNT(*) AS n_rentals,
	COUNT (DISTINCT r.movie_id) AS n_movies,
	AVG (rating) AS avg_rating
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
	WHERE r.movie_id IN (
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01'
GROUP BY ROLLUP (m.year_of_release, c.country)
ORDER BY c.country, m.year_of_release;

country  year_of_release  n_rentals  n_movies  avg_rating
0    Austria           2002.0          1         1    7.000000
1    Austria           2004.0          1         1         NaN
2    Austria           2007.0          1         1    6.000000
3    Austria           2011.0          1         1    6.000000
4    Belgium           2001.0          5         3   10.000000
..       ...              ...        ...       ...         ...
126     None           2012.0         28         5    8.111111
127     None           2013.0         10         2    7.600000
128     None           2014.0          1         1   10.000000
129     None           2014.0          5         1    8.000000
130     None              NaN        333        50    7.902439

[131 rows x 5 columns]

We can see the total aggregation with 333 movie rentals, 50 different movies and an average rating of 7.9. 

Intuitively we can already see that there is no clear pattern of better ratings for movies with later year of release.

## Customer preference by gender

We just saw that customers have no clear preference for more recent movies over older ones. Now the management considers investing money in movies of the best rated genres.

In [17]:
SELECT genre,
	   AVG(rating) AS avg_rating,
	   COUNT(rating) AS n_rating,
       COUNT(*) AS n_rentals,     
	   COUNT(DISTINCT m.movie_id) AS n_movies 
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3 )
AND r.date_renting >= '2018-01-01'
GROUP BY genre
ORDER BY avg_rating DESC; 

genre  avg_rating  n_rating  n_rentals  n_movies
0         Action & Adventure    8.714286         7          9         2
1  Art House & International    8.500000         4          5         1
2                      Other    8.428571         7         16         2
3  Science Fiction & Fantasy    8.276596        47         70        10
4                     Comedy    7.950000        20         31         5
5                  Animation    7.833333         6         10         2
6                      Drama    7.748252       143        245        34
7         Mystery & Suspense    7.428571         7         19         3

Action & Adventure has the highest rating.

## Customer preference for actors

In [18]:
SELECT a.nationality,
       a.gender,
	   AVG(r.rating) AS avg_rating,
	   COUNT(r.rating) AS n_rating,
	   COUNT(*) AS n_rentals,
	   COUNT(DISTINCT a.actor_id) AS n_actors
FROM movies.renting AS r
LEFT JOIN movies.actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN movies.actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01'
GROUP BY CUBE (a.nationality,a.gender) 

nationality  gender  avg_rating  n_rating  n_rentals  n_actors
0          Argentina    male    8.500000         4          5         1
1          Argentina    None    8.500000         4          5         1
2          Australia  female    8.666667         3          5         1
3          Australia    male    7.454545        11         17         3
4          Australia    None    7.714286        14         22         4
5            Austria    male    8.500000         2          6         1
6            Austria    None    8.500000         2          6         1
7            British  female    7.833333        54         78         3
8            British    male    8.105263       114        175         9
9            British    None    8.017857       168        253        12
10            Canada  female    8.000000         5          5         1
11            Canada    male    8.428571         7         13         2
12            Canada    None    8.250000        12         18         3
13           Ireland    male    7.714286         7         16         2
14           Ireland    None    7.714286         7         16         2
15            Israel  female    8.250000         4          8         1
16            Israel    None    8.250000         4          8         1
17             Italy  female    9.000000         4          4         1
18             Italy    None    9.000000         4          4         1
19  Northern Ireland    male    9.000000         4          4         1
20  Northern Ireland    None    9.000000         4          4         1
21       Puerto Rico    male    7.200000         5          7         1
22       Puerto Rico    None    7.200000         5          7         1
23           Somalia    male    7.000000         3          5         1
24           Somalia    None    7.000000         3          5         1
25      South Africa  female    8.214286        14         21         1
26      South Africa    None    8.214286        14         21         1
27             Spain  female    8.500000         4          5         1
28             Spain    male    8.857143         7         12         2
29             Spain    None    8.727273        11         17         3
30               USA  female    7.545455       121        212        27
31               USA    male    7.913223       242        401        42
32               USA    None    7.790634       363        613        69
33              None    None    7.902439       615        999       101
34              None  female    7.751196       209        338        36
35              None    male    7.980296       406        661        65